microsetta_public_api/resources_alt.py

```diff
@@ -0,0 +1,89 @@
+from typing import Dict
+
+from microsetta_public_api.backend import constructors
+from microsetta_public_api.config import (
+    SERVER_CONFIG,
+)
+
+
+class Component:
+
+    def __init__(self, name):
+        self.name = name
+        self.data = None
+        self.children: Dict[str, 'Component'] = dict()
+
+    def add_child(self, child: 'Component'):
+        self.children.update({child.name: child})
+        return child
+
+    def set(self, other: 'Component'):
+        self.name = other.name
+        self.data = other.data
+        self.children = other.children
+
+    def set_data(self, data: object):
+        self.data = data
+
+    def get_data(self):
+        return self.data
+
+    def get_child(self, name: str) -> 'Component':
+        try:
+            return self.children[name]
+        except KeyError:
+            raise KeyError(name)
+
+    def gets(self, *args):
+        if len(args) == 0:
+            return self
+        first = args[0]
+        rest = args[1:]
+        child = self.get_child(first)
+        return child.gets(*rest)
+
+    def has(self, *args):
+        if len(args) == 0:
```

Comment:
I'm trying to grapple with the logic here. If `len(args) == 0` then this instance of a `Component` "has" the passed in arguments. However, if `args` has content, and if the first index of args is not in this `Component`s `children` then it does not "have" the arguments? I'm not sure I follow 

I think it would be helpful for this and `gets` to have some surrounding documentation

Reply:
Might help to clarify that `has` is used to determine if the component has a path starting at `self` in it with names corresponding to the entries of `args`. E.g., for
```
comp = Component('foo')
child = comp.add_child(Component('bar'))
grandchild1 = child.add_child(Component('baz'))
grandchild2 = child.add_child(Component('corge'))
``` 
should return `True` for
`comp.has()`, `comp.has('bar')`, and `comp.has('bar', 'baz')`, `comp.has('bar', 'corge')` and `False` for anyting else.

I will add these explanations to the documentation.

Basically, some call like
```
comp.has('a', 'path', 'of', 'nodes')
```

will give a tuple of args. We grab the first thing off, so in this case `args[0] == 'a'`. Then, we see if the child `'a'` has a path with nodes `('path', 'of', 'nodes')`.
If `a` does not have a child named `path`, then we return `False`, otherwise, we pare down the path again and delegate recursively. This works since the sub-path has to exist in order for the full path to exist. 

If we get to the point where some node was passed `node.has(*empty_tuple)`, then everything in the path before existed, so we can return `True`.